pkg/controller/dns/resource_record_set.go

```diff
@@ -0,0 +1,252 @@
+	if !cmp.Equal(currentSpec, &cr.Spec.ForProvider) {
+		if err := e.kube.Update(ctx, cr); err != nil {
+			return managed.ExternalObservation{}, errors.Wrap(err, errManagedUpdateFailed)
+		}
+		lateInit = true
+	}
+
+	cr.Status.AtProvider = rrsClient.GenerateObservation(*rrs)
+	cr.SetConditions(xpv1.Available())
+
+	upToDate, err := rrsClient.IsUpToDate(meta.GetExternalName(cr), &cr.Spec.ForProvider, rrs)
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, errCheckUpToDate)
+	}
+
+	return managed.ExternalObservation{
+		ResourceExists:          true,
+		ResourceUpToDate:        upToDate,
+		ResourceLateInitialized: lateInit,
+	}, nil
+}
+
+func (e *external) Create(ctx context.Context, mg resource.Managed) (managed.ExternalCreation, error) {
+	cr, ok := mg.(*v1alpha1.ResourceRecordSet)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errNotResourceRecordSet)
+	}
+
+	args := &dns.ResourceRecordSet{}
+	rrsClient.GenerateResourceRecordSet(
+		meta.GetExternalName(cr),
+		cr.Spec.ForProvider,
+		args,
+	)
+
+	rrs, err := e.dns.Create(
+		e.projectID,
+		cr.Spec.ForProvider.ManagedZone,
+		args,
+	).Context(ctx).Do()
+	if err != nil {
+		return managed.ExternalCreation{}, errors.Wrap(err, errCannotCreate)
+	}
+
+	cr.Status.AtProvider = rrsClient.GenerateObservation(*rrs)
+	cr.SetConditions(xpv1.Creating())
+
+	return managed.ExternalCreation{}, nil
+}
+
+func (e *external) Update(ctx context.Context, mg resource.Managed) (managed.ExternalUpdate, error) {
+	cr, ok := mg.(*v1alpha1.ResourceRecordSet)
+	if !ok {
+		return managed.ExternalUpdate{}, errors.New(errNotResourceRecordSet)
+	}
+
+	args := &dns.ResourceRecordSet{}
+	rrsClient.GenerateResourceRecordSet(meta.GetExternalName(cr), cr.Spec.ForProvider, args)
+
+	rrs, err := e.dns.Patch(
+		e.projectID,
+		cr.Spec.ForProvider.ManagedZone,
+		cr.Status.AtProvider.Name,
+		cr.Status.AtProvider.Type,
+		args,
+	).Context(ctx).Do()
+	if err != nil {
+		return managed.ExternalUpdate{}, err
+	}
+
+	cr.Status.AtProvider = rrsClient.GenerateObservation(*rrs)
+
+	return managed.ExternalUpdate{}, nil
+}
+
+func (e *external) Delete(ctx context.Context, mg resource.Managed) error {
+	cr, ok := mg.(*v1alpha1.ResourceRecordSet)
+	if !ok {
+		return errors.New(errNotResourceRecordSet)
+	}
+
+	_, err := e.dns.Delete(
+		e.projectID,
+		cr.Spec.ForProvider.ManagedZone,
+		meta.GetExternalName(cr),
+		cr.Spec.ForProvider.Type,
+	).Context(ctx).Do()
+	if gcp.IsErrorNotFound(err) {
+		return nil
+	}
+
+	cr.SetConditions(xpv1.Deleting())
+
+	return errors.Wrap(err, errCannotDelete)
+}
+
+func (e *external) Get(ctx context.Context, cr *v1alpha1.ResourceRecordSet) (*dns.ResourceRecordSet, error) {
+	rrs, err := e.dns.Get(
+		e.projectID,
+		cr.Spec.ForProvider.ManagedZone,
+		meta.GetExternalName(cr),
+		cr.Spec.ForProvider.Type,
+	).Context(ctx).Do()
+	if err == nil {
+		return rrs, nil
+	} else if !gcp.IsErrorNotFound(err) {
+		return nil, err
+	}
+	if cr.Status.AtProvider.Name != "" {
```

Comment:
same here, let's use external-name instead of relying on `Status`

Reply:
Also confused a bit with the logic here, do we expect `cr.Status.AtProvider.Name` might differ from `meta.GetExternalName(cr)`?

Reply:
Yes, it might differ in the first reconcile loop **after** the user applies with a new external name. I'm using this `cr.Status.AtProvider.Name` to keep track of the last name of the RRS, so I can find it when the user wants to update it.

Reply:
See my comment [above](https://github.com/crossplane/provider-gcp/pull/358#discussion_r695251661)